### PR TITLE
Check $__rate_interval only for rate functions

### PR DESCRIPTION
**Pull request by wuzuf**
Hello,
This should cover a case I have internally, with a query like `sum(increase(foo{...}[$__range]))` interval, which I believe should not be rejected by the linter.
Let me know if you have concerns.
Gabriel

### CODE REVIEW

File: lint/results.go [hunk 1]
```diff
@@ -53,7 +58,7 @@ func (rs *ResultSet) Configure(c *ConfigurationFile) {
 // AddResult adds a result to the ResultSet, applying the current configuration if set
 func (rs *ResultSet) AddResult(r ResultContext) {
 	if rs.config != nil {
-	        r = rs.config.Apply(r)
+		r = rs.config.Apply(r)
```
Comment (tomwilkie) [inline lint/results.go:61]: Is the github lint action not catching this kind of thing?

File: lint/rule_panel_rate_interval.go [hunk 2]
```diff
@@ -2,50 +2,70 @@ package lint
...
+				return SuccessResult
 			}
 
 			if !panelHasQueries(p) {
 				// Don't lint certain types of panels.
-				return Result{
-					Severity: Success,
-					Message:  "OK",
-				}
+				return SuccessResult
```
Comment (tomwilkie) [inline lint/rule_panel_rate_interval.go:38]: I like this, but in the future refactors like this should ideally be a separate PR.

File: lint/rule_panel_rate_interval.go [hunk 3]
```diff
@@ -2,50 +2,70 @@ package lint
...
+				expr, err := parsePromQL(target.Expr, d.Templating.List)
+				if err != nil {
+					continue
+				}
+				err = parser.Walk(inspector(func(node parser.Node, parents []parser.Node) error {
+					selector, ok := node.(*parser.MatrixSelector)
+					if !ok || selector.Range == rateIntervalMagicDuration || len(parents) == 0 {
+						// We are not inspecting something like foo{...}[...] or the range vector selector is $__rate_interval
+						return nil
+					}
```
Comment (tomwilkie) [inline lint/rule_panel_rate_interval.go:51]: I think I'd prefer slightly more verbose if statements; putting 3 different clauses in gets confusing for me :-)

```suggestion
					selector, ok := node.(*parser.MatrixSelector)
					if !ok {
						// We are not inspecting something like foo{...}[...]
						continue
					}
					
					 
					if selector.Range == rateIntervalMagicDuration {
						 // Range vector selector is $__rate_interval
						 return nil
					}
					 
					 if len(parents) == 0 {
						// Bit weird to have a naked foo[$__rate_interval], but allow it.
						return nil
					}
```

**Review by tomwilkie — APPROVED** [review]
Thanks @wuzuf! Mostly LGTM I would just like the if statements simplified a little.

**Comment by wuzuf** [comment]
@tomwilkie any chance I can get that in soon? In currently have a locally patched version but would love to see this fix upstream :)

**Comment by rgeyer** [comment]
> @tomwilkie any chance I can get that in soon? In currently have a locally patched version but would love to see this fix upstream :)

Sorry @wuzuf, this managed to get stale as neither Tom nor I had the bandwidth to keep up.

I'm coming back to maintain a few things here, and merged this PR.

Thanks for the contribution!

**Pull request merged**